lib/utils/util.js

```diff
@@ -301,3 +337,136 @@ function (txJSON, finishCache, nameHashMap, processedFinishes) {
+ *
+ * @param {Output} output
+ * @param {Number} outputIndex
+ * @param {String} txHash
+ * @param {String} idempotencyKey
+ * @returns {ProcessedOutput}
+ */
+
+util.postProcessOutput = function (output,
+  outputIndex,
+  txHash,
+  idempotencyKey) {
+  return {
+    idempotency_key: idempotencyKey,
+    tx_hash: txHash,
+    output_index: outputIndex,
+    output: output,
+    from_cache: false
+  };
+};
+
+/**
+ * Retrieve keys from cache and apply placer mapper
+ * function to each returned cache value and store placers
+ * output in cacheHits.
+ *
+ * @param {LRU} cache processedOutputCache
+ * @param {Array<String>} keys idempotencyKeys to look for within cache
+ * @param {function} placer mapper function to map each cached
+ * value to requested format
+ * @returns {{ cacheHits: Array<ProcessedOutput>, cacheMisses: Array<String> }}
+ * CacheSearchResult
+ */
+util.retrieveFromCache = function (cache, keys, placer) {
+  const cacheMisses = [];
+  const cacheHits = [];
+  keys.forEach((key) => {
+    if (cache.has(key)) {
+      const retrieved = cache.get(key);
+      placer(cacheHits, retrieved);
+    } else {
+      cacheMisses.push(key);
+    }
+  });
+
+  return {
+    cacheMisses,
+    cacheHits
+  };
+};
+
+/**
+ * Retrieves keys from cache, each value is threated as an Array
+ * and results accumulated to cacheHits
+ *
+ * @param {LRU} cache processedOutputCache
+ * @param {Array<String>} keys idempotencyKeys to look for within cache
+ * @returns {{ cacheHits: Array<ProcessedOutput>, cacheMisses: Array<String> }}
+ * CacheSearchResult
+ */
+
+util.retrieveMultipleFromCache = function (cache, keys) {
+  return this.retrieveFromCache(cache, keys, (cacheHits, retrieved) =>
+    retrieved.forEach(element => cacheHits.push(element))
+  );
+};
+
+/**
+ * Retrieves keys from cache
+ *
+ * @param {LRU} cache processedOutputCache
+ * @param {Array<string>} keys idempotencyKeys to look for within cache
+ * @returns {{ cacheHits: Array<ProcessedOutput>, cacheMisses: Array<String> }}
+ * CacheSearchResult
+ */
+
+util.retrieveSingleFromCache = function (cache, keys) {
+  return this.retrieveFromCache(cache, keys, (cacheHits, retrieved) =>
+    cacheHits.push(retrieved)
+  );
+};
+
+/**
+ * Stores processedOutput in cache, using the placer
+ * to generate proper key
+ *
+ * @param {ProcessedOutput} processedOutput asd
+ * @param {LRU} cache asd
+ * @param {function} placer asd
+ * @returns {undefined}
+ */
+util.storeInCache = function (processedOutput, cache, placer) {
+  const cachedItem = Object.assign({}, processedOutput);
+  cachedItem.from_cache = true;
+  placer(cache, cachedItem);
+};
+
+/**
+ * Stores processedOutput in cache as a member of an array
+ *
+ * @param {ProcessedOutput} processedOutput 
+ * @param {LRU} cache processedOutputCache
+ * @returns {undefined}
+ */
+util.storeMultipleInCache = function (processedOutput, cache) {
+  this.storeInCache(processedOutput, cache, (cache, processedOutput) => {
```

Comment:
nit: in all these helpers, it would be preferable to use different variable names for inner scoped values instead of relying on scoping rules. There are two senses of `cache` and `processedOutput` here so it's difficult to discern the intent of the code

Reply:
nice feedback will definitely keep in mind

Reply:
fixing

Reply:
fixed